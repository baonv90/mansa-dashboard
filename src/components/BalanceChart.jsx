import React from 'react'
import styled from 'styled-components';
import Trend from 'react-trend';

const Container = styled.div`
  padding: .5rem;
  border-radius: 10px;
  background: var(--grey-color);
  margin-top: 1.5rem;
`;

const Chart = styled(Trend)`
  max-width: 100%;
`;

// Component to display chart related to transactions
const BalanceChart = ({ data = [] }) => (
  <Container>
    <Chart
      data={data}
      smooth
      gradient={['var(--out-color)', 'var(--in-color)']}
      radius={10}
      autoDraw
      autoDrawDuration={1500}
      autoDrawEasing="ease-in"
      strokeWidth={3}
      strokeLinecap={'round'}
      width={300}
      height={200}
    />
  </Container>
);

export default BalanceChart;