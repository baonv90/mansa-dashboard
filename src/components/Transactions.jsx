import React from 'react';
import styled from 'styled-components';


const Container = styled.div`
  display: flex;
  flex-direction: column;
  overflow: auto;
`;

export const Item = styled.div`
  display: flex;
  font-size: .825rem;
  line-height: 1.25rem;
  align-items: center;
  padding: .5rem;
  margin-right: 1rem;
  border-bottom: 1px solid var(--grey-color);
`;

const Header = styled(Item)`
  font-weight: 600;
  border-bottom: 1px solid var(--border-color);
`;

const Info = styled.div`
  width: 25%;
`;

const Amount = styled.div`
  width: 25%;
  display: flex;
  color: ${({ moneyIn, header }) => header 
    ? 'var(--title-color)' 
    : moneyIn ? 'var(--in-color)' : 'var(--out-color)'};
  justify-content: flex-end;
  font-weight: 600;
`;

// Simple text format
const formatText = (str) => {
  return str && str.replace('_', ' ').toLowerCase();
};

// Transaction history
const Transactions = ({ data }) => {
  return (
    <>
      <Header>
        <Info>Date</Info>
        <Info>Type</Info>
        <Info>Category</Info>
        <Amount header={true}>Amount</Amount>
      </Header>
      <Container>
        {(data || []).map((dat, index) => {
          const { timestamp, amount, currency, transaction_category, transaction_type } = dat;
          const displayDate = timestamp.slice(0, timestamp.indexOf(' '));

          return (
            <Item key={`transaction-${index}`}>
              <Info>{displayDate}</Info>
              <Info>{formatText(transaction_type)}</Info>
              <Info>{formatText(transaction_category)}</Info>
              <Amount moneyIn={amount >= 0}>{amount}{currency === 'GBP' ? 'Â£' : currency}</Amount>
            </Item>
          );
        })}
      </Container>
    </>
  );
};

export default Transactions;