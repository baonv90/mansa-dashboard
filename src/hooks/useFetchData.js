import { useState, useEffect } from 'react';

// A hook to fetch data from given url
// it returns the fetching status loading and data
export const useFetchData = (url) => {
  const [data, setData] = useState();
  const [loading, setLoading] = useState(false);
  const [hasError, setHasError] = useState(false);

  useEffect(() => {
    setLoading(true);
    fetch(url).then((res) => {
      res.json().then((data) => {
        setData(data);
        setLoading(false);
      })
      .catch(() => {
        setHasError(true);
        setLoading(false);
      });
    });
  }, [url]);

  return { data, loading, hasError };
};