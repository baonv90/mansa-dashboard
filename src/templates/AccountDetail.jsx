import React from 'react'
import styled from 'styled-components';

import Block from '../layout/Block';
import BalanceChart from '../components/BalanceChart';
import Transactions from '../components/Transactions';
import { useFetchData } from '../hooks/useFetchData';
import { data_source } from './AccountInfo';

// For testing purpose, the start and end dates are fixed
const from = new Date('01 January 2019 00:00 UTC').toISOString();
const to = new Date('01 June 2019 00:00 UTC').toISOString();

const Container = styled.div`
  display: flex;
  margin-top: 1.5rem;
  max-height: 60vh;
`;

const BalanceWrapper = styled.div`
  display: flex;
  flex-direction: column;
  width: 35%;
`;

const TransactionsWrapper = styled.div`
  display: flex;
  flex: 1;
  flex-direction: column;
  margin-left: 1rem;
`;

const Balance = styled.div`
  display: flex;
  align-items: baseline;
  margin-top: 1rem;
`;

const AccountBalance = styled.span`
  font-size: 2rem;
  line-height: 1.5rem;
  font-weight: 800;
`;

const Currency = styled.span`
  font-size: .8rem;
  line-height: .8rem;
  font-weight: 400;
  margin-left: .25rem;
`;

// Component to display current balance, balance chart, transaction history of one account
const AccountDetail = (props) => {
  const { available, currency, account_id } = props;
  const { data } = useFetchData(data_source + '/' + account_id + '/transactions?from=' + from + '&to=' + to);

  if (!data) {
    return null;
  }

  const chartData = (data || []).map(dat => dat.amount);

  return (
    <Container>
      <BalanceWrapper>
        <Block title='Current Balance'>
          <Balance>
            <AccountBalance>{available}</AccountBalance>
            <Currency>{currency}</Currency>
          </Balance>
        </Block>
        {chartData && chartData.length && <BalanceChart data={chartData} />}
      </BalanceWrapper>
      <TransactionsWrapper>
        <Block title='Transactions'>
          <Transactions data={data || []}/>
        </Block>
      </TransactionsWrapper>
    </Container>
  );
};

export default AccountDetail;
