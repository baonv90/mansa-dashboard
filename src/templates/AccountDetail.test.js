import renderer from 'react-test-renderer';
import AccountDetail from './AccountDetail';
import Transactions from '../components/Transactions';
import BalanceChart from '../components/BalanceChart';

import { useFetchData } from '../hooks/useFetchData';

jest.mock('../hooks/useFetchData');
jest.mock('../components/BalanceChart')

describe('templates > AccountDetail', () => {
  const fakeData = [{
    timestamp: '10/10/2021',
    amount: 10,
    transaction_type: 'DEBIT',
    transaction_category: 'DEBIT'
  }];

  beforeEach(() => {
    useFetchData.mockReturnValue({
      loading: false,
      data: fakeData
    });

    BalanceChart.mockReturnValue(() => <div>chart</div>);
  });

  it('should match snapshot', async () =>  {
    const component = await renderer.create(
      <AccountDetail />,
    );
    
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should pass data correctly to Transactions', async () =>  {
    const tree = await renderer.create(
      <AccountDetail />,
    );
    
    const transactions = tree.root.findByType(Transactions);
    expect(transactions.props.data).toEqual(fakeData);
  });
});
