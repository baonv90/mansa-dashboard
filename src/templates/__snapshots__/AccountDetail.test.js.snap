// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`templates > AccountDetail should match snapshot 1`] = `
<div
  className="sc-iqAclL cuMlls"
>
  <div
    className="sc-crzoAE fuFIuo"
  >
    <span
      className="sc-bdnxRM bnAtid"
    >
      Current Balance
    </span>
    <div
      className="sc-bqGGPW knRyLe"
    >
      <span
        className="sc-ksluID jtYqrw"
      />
      <span
        className="sc-hBMUJo jFxRFz"
      />
    </div>
  </div>
  <div
    className="sc-dIsUp eIXnfA"
  >
    <span
      className="sc-bdnxRM bnAtid"
    >
      Transactions
    </span>
    <div
      className="sc-dlnjwi sc-hKFxyN bKaicu jJSnIS"
    >
      <div
        className="sc-eCApnc QOfmb"
      >
        Date
      </div>
      <div
        className="sc-eCApnc QOfmb"
      >
        Type
      </div>
      <div
        className="sc-eCApnc QOfmb"
      >
        Category
      </div>
      <div
        className="sc-jSFjdj erYJoO"
      >
        Amount
      </div>
    </div>
    <div
      className="sc-gtsrHT eCTvhQ"
    >
      <div
        className="sc-dlnjwi bKaicu"
      >
        <div
          className="sc-eCApnc QOfmb"
        >
          10/10/202
        </div>
        <div
          className="sc-eCApnc QOfmb"
        >
          debit
        </div>
        <div
          className="sc-eCApnc QOfmb"
        >
          debit
        </div>
        <div
          className="sc-jSFjdj fJpqKh"
        >
          10
        </div>
      </div>
    </div>
  </div>
</div>
`;
